packages/neon_framework/lib/src/utils/request_manager.dart

```diff
@@ -137,24 +146,44 @@ class RequestManager {
     required UnwrapCallback<T, R> unwrap,
     required SerializeCallback<R> serialize,
     required DeserializeCallback<R> deserialize,
+    GetCacheParameters? getCacheParameters,
     bool disableTimeout = false,
     Duration timeLimit = kDefaultTimeout,
     int retries = 0,
   }) async {
     final key = '${account.id}-$cacheKey';
 
-    Future<void>? cacheFuture;
     if (retries == 0) {
       // emit loading state with the current value if present
       final value = subject.valueOrNull?.asLoading() ?? Result.loading();
       subject.add(value);
 
-      cacheFuture = _emitCached(
+      final cachedValue = await _emitCached(
         key,
         subject,
         unwrap,
         deserialize,
       );
+
+      if (cachedValue != null && cachedValue.value != null) {
+        if (cachedValue.expires?.isAfter(DateTime.now()) ?? false) {
+          subject.add(subject.value.copyWith(isLoading: false));
+          return;
+        }
+
+        final cacheParameters = await getCacheParameters?.call();
+        if (cacheParameters?.etag != null && cacheParameters!.etag == cachedValue.etag) {
+          unawaited(
+            _cache?.updateParameters(
+              key,
+              cacheParameters.etag,
+              cacheParameters.expires != null ? _parseExpires(cacheParameters.expires!) : null,
+            ),
+          );
+          subject.add(subject.value.copyWith(isLoading: false));
+          return;
+        }
+      }
     }
```

Comment:
There is a lot of cache handling going on here.
Now that we only ever call `emitCache` once we should think about merging it into the main wrap method.

This would probably allow us easier logic cleanups and more readability.

Reply:
I agree, but I think we only called the method once before this already :thinking: